org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/InlayHintVisitor.java

```diff
@@ -123,6 +128,60 @@ public boolean visit(SuperConstructorInvocation node) {
 		return true;
 	}
 
+	@Override
+	public boolean visit(LambdaExpression node) {
+		if (isOutOfRange(node) || isGenerated(node) || !isParameterTypeHintsEnabled) {
+			return true;
+		}
+
+		// Get the method binding for the lambda to understand parameter types
+		IMethodBinding methodBinding = node.resolveMethodBinding();
+		if (methodBinding == null) {
+			return true;
+		}
+
+		// Get parameter types from the functional interface
+		ITypeBinding[] parameterTypes = methodBinding.getParameterTypes();
+		List<?> parameters = node.parameters();
+
+		if (parameters.size() != parameterTypes.length) {
+			return true;
+		}
+
+		try {
+			for (int i = 0; i < parameters.size(); i++) {
+				Object param = parameters.get(i);
+				if (param instanceof SingleVariableDeclaration) {
```

Comment:
If the programmer uses a mix of implicit and explicit parameter types, the code won't compile or even be parsed correctly. So if the first entry is a SingleVariableDeclaration, I think it's safe to skip trying the rest of the entries. This could speed things up a bit, but in practice lambdas usually only have 1 or 2 parameters, so I don't think it'll be by much.

Reply:
indeed. I make it return early now.